cmake/Platform/Generation/ArduinoFirmwareGenerator.cmake

```diff
@@ -10,6 +10,8 @@ function(GENERATE_ARDUINO_FIRMWARE INPUT_NAME)
             "BOARD;BOARD_CPU;PORT;SKETCH;PROGRAMMER" # One Value Keywords
             "SERIAL;SRCS;HDRS;LIBS;ARDLIBS;AFLAGS"   # Multi Value Keywords
             ${ARGN})
+            
+    include(FixRedundantTargetCompileFlags)
```

Comment:
This is definitely not a good place to include a cmake script.
What will happen is that inside the running `generate_arduino_firmware` function, you'll define another function and then eventually use it. Not only that is strange, I'm surprised it has even worked!
Generally, cmake `include`s should be only at the very first cmake script being processed, in our case - It's `Arduino.cmake`. Although it seems correct to include stuff only where its' relevant, as you would do in any programming language, cmake behaves a bit differently. 
If I'm somehow wrong, please correct me 😄

Reply:
That's a thing that happened to me a long time ago. When I started working with CMake, I thought it is a proper programming language. But it is __definitely not__. 

While I doubt it is turing-complete, it definitely does not feature proper scoping. There is scoping for variables in functions but that's all. A function can appear anywhere an it is visible __globally__ once defined. A function always silently overwrites a  predefined function with same name and a locally defined function overrides a predefined global one. I never quite figured out the exact rules (if there are any) :cry:
Functions and scopes are definitely one of the worst aspects of CMake.

Just look at the following snipped, but brace yourself... Just by looking at it, you will never be able to find out what it actually does. Use it as script and call with `cmake -P` to find out.
```cmake
function(A)
   message("A1")
endfunction()

function(A)
   message("A2")

   function(A)
      message("A3")
   endfunction()

   A()
   
   function(B)
      message("B")
   endfunction()
endfunction()

A()
B()
```

> This is definitely not a good place to include a cmake script.

You use the term CMake script in a slightly peculiar way when you say that I include a script. In CMake's own terminology, a script is something that is supposed to be called using `cmake -P <script.cmake>`. This is then executed in *script mode* and there are specific rules about what CMake functions are allowed to be called in this mode of execution.
I would prefer to speak of including CMake code or similar, trying to avoid the word *script* in the context of includes. 

> Generally, cmake includes should be only at the very first cmake script being processed, in our case - It's Arduino.cmake. Although it seems correct to include stuff only where its' relevant, as you would do in any programming language, cmake behaves a bit differently.
If I'm somehow wrong, please correct me

There's no such convention.
I am working on a huge simulation software with a build system that consists of several hundred files of CMake code that are nested in all sorts of ways. You can only keep track of such a *monster* when you use proper including in a very similar fashion as you would apply e.g. in C or as you import modules in Python (there it is common to have import directives in function scope). Just imagine a C program where you would only be allowed to include headers in `main.c` :wink:

It is perfectly valid to add includes of *exotic* functionality close to where it is used to emphasis on the relation to where they are called. This is especially important when CMake files grow larger. Due to CMake's bad scoping rules it is not always easy to divide them up again (you probably experienced that :smile:) and having the include directives close can reduce the time it takes to search for the file where a function is defined.
If you are afraid of additional overhead, there is BTW a slight performance gain when using [include guards](https://cmake.org/cmake/help/latest/command/include_guard.html). Using include guards, there is no measurable performance loss from including CMake code files several times.

One thing is important, though. One must distinguish CMake files that contain function and macro definitions only and those that also contain code outside of function scope. The former can safely be included anywhere, while the latter must be handled with care.
I regard it safest to define anything inside a function or macro, even stuff that needs to be executed at a parent/global scope (this is actually the only remaining use of CMake macros).
This avoids problems with including. It is better to have macros like
```
macro(init ...)
...
endmacro()
```
in an included file that are explicitly called at a suitable place instead of having the same code executed accidentally when (multiply) including a file.

As this is not always possible e.g. when confronted with legacy code I use the following file name convention: `foo.macros.cmake` to name files that only contain functions and macro definitions and `bar.cmake` for the rest, except for CMake files that are supposed to be used in script mode. Those I name `foo.script.cmake`. Files ending with `.macros.cmake` can be safely included anywhere. When including, I always try to use the full (relative) path and filename of the included file. This is something that I am missing in Arduino-CMake and that might be annoying to users in some places. It would make searching for included files much easier.

Please excuse, this is ended up as an awful lot of text. But I thought it worthwhile sharing some of the 'wisdom' I painfully gathered in the last fifteen years working with CMake. I still like CMake for it replaced those awful autotools but if I could, I would make it a lot more consistent and easier to work with.

Reply:
Well, that's sure as hell a lot to process, but it's definitely worth it - Thank you for sharing your knowledge 😃 

I get all of your points, and since you're **much** more experienced in cmake than I am, I have no other option but agreeing lol
**However**, I still don't like the include of the cmake file *inside* the function. It is sometimes common in a dynamic language such as python, but in C on the contrary (Or any other statically typed language), not so much. 
Wouldn't it be better to include the function at the top of the function at least, if not at the top of the whole file itself (If I understood you correctly it *may* cause problems with the functions we're trying to include)?

> It is perfectly valid to add includes of exotic functionality close to where it is used to emphasis on the relation to where they are called. 

You've said it yourself. However, in **reality** you call the function a few lines below the include. (Note the difference here between *few* and *a few* 😄). So what am I missing?

Reply:
I sometimes include close to the line of use but mostly include files that contain just one function that is of no use in other places. In other (most) occasions I include at the top of the file where a function is called. 

As there might be other users that are also confused by includes inside of functions, lets stick with the convention of exclusively including at the top of CMake files. I will update this PR accordingly.

Reply:
Excellent, thanks 👍